lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs

```diff
@@ -15,28 +17,57 @@ module Cardano.Wallet.Primitive.AddressDiscovery.Random
     (
     -- ** State
       RndState (..)
-    , mkRndState
     ) where
 
 import Prelude
 
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..), Key (..), XPrv )
+import Cardano.Wallet.Primitive.AddressDerivation.Random
+    ( decryptDerPath, eskToHdPassphrase )
 import Cardano.Wallet.Primitive.AddressDiscovery
     ( CompareDiscovery (..)
     , GenChange (..)
     , IsOurs (..)
     , IsOwned (..)
     , KnownAddresses (..)
     )
-import Data.ByteString
-    ( ByteString )
+import Control.DeepSeq
+    ( NFData )
+import Crypto.Error
+    ( CryptoFailable (..) )
+import Data.Map.Strict
+    ( Map )
+import Data.Text
+    ( Text )
+import GHC.Generics
+    ( Generic )
 
-newtype RndState = RndState ByteString
+import qualified Data.Map.Strict as Map
 
-mkRndState :: RndState
-mkRndState = error "AddressDiscovery.Random unimplemented"
+-- | Unified identifier for wallets (both FO and EO ones).
+-- Technically it contains a text, but actually it contains one from two options:
+--
+-- 1. for FO-wallets - 'Address' made from wallet's root public key, in Base58-format.
+-- 2. for EO-wallets - 'UUID', in Base58-format.
+newtype HdRootId = HdRootId { getHdRootId :: Text }
+    deriving (Eq, Ord, Show, Generic)
+
+instance NFData HdRootId
+
+newtype RndState = RndState { getRndState :: Map HdRootId (Key 'RootK XPrv) }
+    deriving (Generic)
+
+instance NFData RndState
 
 instance IsOurs RndState where
-    isOurs _ s = (False, s)
+    isOurs addr (RndState s) =
+        (,RndState s) $ Map.foldl' (||) False $ flip Map.mapWithKey s $ \_ esk -> do
+        case decryptDerPath (eskToHdPassphrase esk) addr of
```

Comment:
I think this would need to
1. Extract the HD payload part of the address
2. Run `decodeAddressDerivationPath` on the payload, using the root key.

Reply:
yes, you are right.